### PR TITLE
validate bpo number

### PR DESCRIPTION
**Pull request by levlaz**
Add helper function to check if the issue number exists on bpo.

set FAILURE_STATUS in the event that the issue does not exist.

Fix https://github.com/python/bedevere/issues/15

### CODE REVIEW

File: bedevere/bpo.py [hunk 1]
```diff
@@ -21,7 +23,7 @@
 SKIP_ISSUE_LABEL = util.skip_label("issue")
 STATUS_CONTEXT = "bedevere/issue-number"
 # Try to keep descriptions at or below 50 characters, else GitHub's CSS will truncate it.
-_FAILURE_DESCRIPTION = 'No issue # in title or "skip issue" label found'
+_FAILURE_DESCRIPTION = 'Issue does not exit, No issue # in title or "skip issue" label found'
```
Comment (levlaz) [inline bedevere/bpo.py]: This felt a little dirty, but I was not sure if we wanted to "yet another" edge case of failure description. Happy to refactor this if we want to be more specific of the reason for failure.

File: tests/test_bpo.py [hunk 2]
```diff
@@ -267,3 +290,11 @@ def __init__(self, *, getitem=None):
     await bpo.router.dispatch(event, gh)
     assert gh.patch_data is None
     assert gh.patch_url is None
+
+
+def test_validate_issue_number():
```
Comment (levlaz) [inline tests/test_bpo.py]: This is a "unit" test but it actually makes a call to bpo... maybe this should be mocked as well (or it may not be necessary all together).

File: bedevere/bpo.py (hunk 1 above)
Comment (Mariatta) [reply bedevere/bpo.py]: Hmm, this might be too long as a status check message.
I'm thinking it can be a separate status check, maybe "bedevere/valid-issue"? @brettcannon, what do you think?

File: bedevere/bpo.py [hunk 3]
```diff
@@ -89,3 +94,13 @@ def create_success_status(found_issue):
     return util.create_status(STATUS_CONTEXT, util.StatusState.SUCCESS,
                               description=f"Issue number {issue_number} found",
                               target_url=url)
+
+def validate_issue_number(issue_number):
+    """Validate that the issue number exists on bpo"""
```
Comment (Mariatta) [inline bedevere/bpo.py]: The docstring should end with a period :)

File: bedevere/bpo.py [hunk 4]
```diff
@@ -89,3 +94,13 @@ def create_success_status(found_issue):
...
+    url = f"https://bugs.python.org/issue{issue_number}"
+    req = urllib.request.Request(url=url, method='HEAD')
+    try:
+        urllib.request.urlopen(req)
+        return True
+    except urllib.error.HTTPError:
+        return False
```
Comment (Mariatta) [inline bedevere/bpo.py]: Needs a newline.

File: tests/test_bpo.py (hunk 2 above)
Comment (Mariatta) [reply tests/test_bpo.py]: My preference is to mock the call to b.p.o.

File: bedevere/bpo.py [hunk 5]
```diff
@@ -1,5 +1,7 @@
 """Check if a bugs.python.org issue number is specified in the pull request's title."""
 import re
+import urllib.request
```
Comment (Mariatta) [inline bedevere/bpo.py]: How about `from urllib import error, request` instead?

**Comment by levlaz** [comment]
Thank you for the review @Mariatta, I made all of the requested changes except for defining a new status check message since you want to get some feedback from Brett. 

Happy to implement whatever you all decide.

File: bedevere/bpo.py (hunk 1 above)
Comment (brettcannon) [reply bedevere/bpo.py]: I don't think we need another status check (we are already at 4). Is there any reason the message can't be dynamically specified like for the `NEWS.d` check? In this instance the current message is the one that shows up if anything is found, but if an issue number is discovered but found not to exist, a different message saying something like "issue #NNNN not found at bugs.python.org" or something would be set while still flagging a status check failure.

Make sense?

Comment (Mariatta) [reply bedevere/bpo.py]: Using a different message sounds good :)

File: bedevere/bpo.py [hunk 6]
```diff
@@ -89,3 +92,13 @@ def create_success_status(found_issue):
     return util.create_status(STATUS_CONTEXT, util.StatusState.SUCCESS,
                               description=f"Issue number {issue_number} found",
                               target_url=url)
+
+def validate_issue_number(issue_number):
```
Comment (brettcannon) [inline bedevere/bpo.py]: Start with an underscore since this shouldn't be used outside of this module.

File: bedevere/bpo.py (hunk 1 above)
Comment (terryjreedy) [reply bedevere/bpo.py]: Corrections to message: /exit/exist/, /No/no/, /title/title,/.  But +1 to situation specific message.   This is not an exception raised in inner loops for control flow.  (The reason for some quick and vague exception messages.)

File: bedevere/bpo.py [hunk 7]
```diff
@@ -89,3 +92,13 @@ def create_success_status(found_issue):
...
+    """Validate that the issue number exists on bpo."""
+    url = f"https://bugs.python.org/issue{issue_number}"
+    req = request.Request(url=url, method='HEAD')
+    try:
+        res = request.urlopen(req)
```
Comment (merwok) [inline bedevere/bpo.py]: Shouldn’t this use an async request instead of dear old urllib?

**Comment by Mariatta** [comment]
@levlaz still working on this?

**Comment by lysnikolaou** [comment]
Since @levlaz is apparently no longer working on this, should this PR maybe a closed? @Mariatta?

**Comment by brettcannon** [comment]
Due to this stalling out, I'm closing the PR.
